### PR TITLE
feat: make the builtin `extract` plugin support compressed tarballs (`*.tar.gz`, `*.tar.bz2`, etc.)

### PR DESCRIPTION
**Pull request by DirkFi**
Close [#1548](https://github.com/sxyazi/yazi/issues/1548).

Description:
 * Add `spawn_tar` function and `is_tar_compatible` function
   - Create `spawn_tar` function to run command with `tar -xaf` to unzip file to tmp directory
   - Create `is_tar_compatible` function to tell if a zip file can be extracted with `tar` command
 * Add `os.excute()` line to make sure the tmp dir exists because `tar` command will not automatically create that dir, unlike `7z`
 * Refactor the logic in the `try()` function to if the zip file can be unzipped with tar and no password required, use `tar`, otherwise use `7z` instead

### CODE REVIEW

File: yazi-plugin/preset/plugins/extract.lua [hunk 1]
```diff
@@ -47,23 +57,45 @@ function M:try_with(url, pwd)
 		fail("Failed to determine a temporary directory for %s", url)
 	end
 
+	os.execute("mkdir -p " .. tostring(tmp))
```
Comment (sxyazi) [inline yazi-plugin/preset/plugins/extract.lua:60]: Please add a new method `fs.create(type, url)` above and similar to the `fs.remove()` method here and use it instead to achieve cross-platform compatibility https://github.com/sxyazi/yazi/blob/640e497c358c5ef576643e646cdb73be67bf7c52/yazi-plugin/src/fs/fs.rs#L36-L54

- type: available values `"file"`, `"dir"`
- url: Url to operate

File: yazi-plugin/preset/plugins/extract.lua [hunk 2]
```diff
@@ -47,23 +57,45 @@ function M:try_with(url, pwd)
...
+			fail("Failed to spawn `tar` command, error code %s", code)
+		end
 
-	local output, err = child:wait_with_output()
-	if output and output.status.code == 2 and archive:is_encrypted(output.stderr) then
-		fs.remove("dir_clean", tmp)
-		return true -- Needs retry
-	end
+		local output, err = child:wait_with_output()
+		if output and output.status.code == 2 and archive:is_encrypted(output.stderr) then
```
Comment (sxyazi) [inline yazi-plugin/preset/plugins/extract.lua:69]: `archive:is_encrypted(output.stderr) `, Can `tar` be encrypted?

File: yazi-plugin/preset/plugins/extract.lua [hunk 3]
```diff
@@ -47,23 +57,45 @@ function M:try_with(url, pwd)
...
 	end
 
+	os.execute("mkdir -p " .. tostring(tmp))
 	local archive = require("archive")
-	local child, code = archive:spawn_7z { "x", "-aou", "-p" .. pwd, "-o" .. tostring(tmp), tostring(url) }
-	if not child then
-		fail("Spawn `7z` and `7zz` both commands failed, error code %s", code)
-	end
+	if self:is_tar_compatible(tostring(url)) then
+		local child, code = archive:spawn_tar { "-xvf", tostring(url), "-C", tostring(tmp) }
```
Comment (sxyazi) [inline yazi-plugin/preset/plugins/extract.lua:63]:
```suggestion
		local child, code = archive:spawn_tar { "-xf", tostring(url), "-C", tostring(tmp) }
```

Any reason for specifying the `-v` switch?

File: yazi-plugin/preset/plugins/extract.lua [hunk 4]
```diff
@@ -36,6 +36,16 @@ function M:entry(args)
 	end
 end
 
+function M:is_tar_compatible(url)
+	local tar_extensions = { ".tar", ".tgz", ".tar.gz", ".tar.bz2" }
```
Comment (sxyazi) [inline yazi-plugin/preset/plugins/extract.lua:40]: Are you sure `tar -xvf` also works for `.tar.gz` and `.tar.bz2` files?

File: yazi-plugin/preset/plugins/archive.lua [hunk 5]
```diff
@@ -53,6 +53,17 @@ function M:seek(units)
 	end
 end
 
+function M:spawn_tar(args)
+	local command = "tar"
```
Comment (sxyazi) [inline yazi-plugin/preset/plugins/archive.lua:57]:
```suggestion
```

**Review by sxyazi — CHANGES_REQUESTED** [review]

File: yazi-plugin/preset/plugins/extract.lua (hunk 2 above)
Comment (DirkFi) [reply yazi-plugin/preset/plugins/extract.lua:69]: Tar does not have any built-in encryption like 7z, needs to be used with gpg or other tools.

File: yazi-plugin/preset/plugins/extract.lua (hunk 3 above)
Comment (DirkFi) [reply yazi-plugin/preset/plugins/extract.lua:63]: Ur right, it's an optional one for printing the output of filename, gonna delete this.

**Comment by DirkFi** [comment]
BTW, Do I need to combine the other commits or I can have separate commit in the same PR?

**Comment by veltza** [comment]
There are more [suffixes](https://en.wikipedia.org/wiki/Tar_(computing)#Suffixes_for_compressed_files) than ".tar", ".tgz", ".tar.gz", ".tar.bz2". And short forms should be supported as well.

**Comment by sxyazi** [comment]
> BTW, Do I need to combine the other commits or I can have separate commit in the same PR?

You can make multiple commits in this PR, and I'll squash and merge them into one.

**Comment by DirkFi** [comment]
> There are more [suffixes](https://en.wikipedia.org/wiki/Tar_(computing)#Suffixes_for_compressed_files) than ".tar", ".tgz", ".tar.gz", ".tar.bz2". And short forms should be supported as well.

Thanks! Definitely will go test and add this!

File: yazi-plugin/preset/plugins/extract.lua (hunk 4 above)
Comment (DirkFi) [reply yazi-plugin/preset/plugins/extract.lua:40]: Yeah, it did work. But think it's still necessary to change to `-a` to make it  `--auto-compress` to increase the compatibility.

**Comment by veltza** [comment]
Is this going to fix the preview issues with tar files as well?

**Comment by DirkFi** [comment]
> Is this going to fix the preview issues with tar files as well?

What issue specifically? If you're talking about preview does not show the final file of the `tar.gz` or other type of `tar`, no. It is not fixed in this PR.

**Comment by veltza** [comment]
> What issue specifically? If you're talking about preview does not show the final file of the tar.gz or other type of tar, no. It is not fixed in this PR.

That's what I was referring to. And I hoped that this would have killed two birds with one stone. But that's okay because I think it will be easy to fix after this PR.

**Comment by sxyazi** [comment]
I finally had time to test it today (sorry for the delay), I found that we can directly reuse 7zip to extract tar files, which eliminates the need for an additional dependency. Plus, 7zip is more cross-platform friendly (some older Windows systems can also use it)

**Review by sxyazi — APPROVED** [review]
Thanks!

**Pull request merged**